registry/handlers/images.go

```diff
@@ -11,178 +11,158 @@ import (
+func GetImageManifest(ctx *Context, w http.ResponseWriter, r *http.Request) *HTTPError {
+	ctxu.GetLogger(ctx).Debug("GetImageManifest")
+	tag, dgst := parseTagDigest(ctx, r)
+	manifests := ctx.Repository.Manifests()
 
 	var (
 		sm  *manifest.SignedManifest
 		err error
 	)
 
-	if imh.Tag != "" {
-		sm, err = manifests.GetByTag(imh.Tag)
+	if tag != "" {
+		sm, err = manifests.GetByTag(tag)
 	} else {
-		sm, err = manifests.Get(imh.Digest)
+		sm, err = manifests.Get(dgst)
 	}
 
 	if err != nil {
-		imh.Errors.Push(v2.ErrorCodeManifestUnknown, err)
-		w.WriteHeader(http.StatusNotFound)
-		return
+		return NewHTTPError(v2.ErrorCodeManifestUnknown, err, http.StatusNotFound)
 	}
 
 	// Get the digest, if we don't already have it.
-	if imh.Digest == "" {
-		dgst, err := digestManifest(imh, sm)
+	if dgst == "" {
+		dgst, err = digestManifest(ctx, sm)
 		if err != nil {
-			imh.Errors.Push(v2.ErrorCodeDigestInvalid, err)
-			w.WriteHeader(http.StatusBadRequest)
-			return
+			return NewHTTPError(v2.ErrorCodeDigestInvalid, err, http.StatusBadRequest)
 		}
 
-		imh.Digest = dgst
 	}
 
 	w.Header().Set("Content-Type", "application/json; charset=utf-8")
 	w.Header().Set("Content-Length", fmt.Sprint(len(sm.Raw)))
-	w.Header().Set("Docker-Content-Digest", imh.Digest.String())
+	w.Header().Set("Docker-Content-Digest", dgst.String())
 	w.Write(sm.Raw)
+	return nil
 }
 
 // PutImageManifest validates and stores and image in the registry.
-func (imh *imageManifestHandler) PutImageManifest(w http.ResponseWriter, r *http.Request) {
-	ctxu.GetLogger(imh).Debug("PutImageManifest")
-	manifests := imh.Repository.Manifests()
+func PutImageManifest(ctx *Context, w http.ResponseWriter, r *http.Request) *HTTPError {
+	ctxu.GetLogger(ctx).Debug("PutImageManifest")
+	tag, sentDgst := parseTagDigest(ctx, r)
+	manifests := ctx.Repository.Manifests()
 	dec := json.NewDecoder(r.Body)
 
 	var manifest manifest.SignedManifest
 	if err := dec.Decode(&manifest); err != nil {
-		imh.Errors.Push(v2.ErrorCodeManifestInvalid, err)
-		w.WriteHeader(http.StatusBadRequest)
-		return
+		return NewHTTPError(v2.ErrorCodeManifestInvalid, err, http.StatusBadRequest)
 	}
 
-	dgst, err := digestManifest(imh, &manifest)
+	dgst, err := digestManifest(ctx, &manifest)
 	if err != nil {
-		imh.Errors.Push(v2.ErrorCodeDigestInvalid, err)
-		w.WriteHeader(http.StatusBadRequest)
-		return
+		return NewHTTPError(v2.ErrorCodeDigestInvalid, err, http.StatusBadRequest)
 	}
 
 	// Validate manifest tag or digest matches payload
-	if imh.Tag != "" {
-		if manifest.Tag != imh.Tag {
-			ctxu.GetLogger(imh).Errorf("invalid tag on manifest payload: %q != %q", manifest.Tag, imh.Tag)
-			imh.Errors.Push(v2.ErrorCodeTagInvalid)
-			w.WriteHeader(http.StatusBadRequest)
-			return
+	if tag != "" {
+		if manifest.Tag != tag {
+			ctxu.GetLogger(ctx).Errorf("invalid tag on manifest payload: %q != %q", manifest.Tag, tag)
+			return NewHTTPError(v2.ErrorCodeTagInvalid, nil, http.StatusBadRequest)
```

Comment:
This is an example of where we are going to get future problems due to a nillable parameter. This should just be this:

```go
var errors v2.Errors // declared near top of every handler
...
		if manifest.Tag != tag {
			ctxu.GetLogger(ctx).Errorf("invalid tag on manifest payload: %q != %q", manifest.Tag, tag)
			return v2.ErrorCodeTagInvalid // note we'll have to make ErrorCode implement error
		}
```

Then, in the calling function, we have a single type switch that is detecting this:

```go
err := handler(...)
if err != nil {
        serveError(w, r, err)
}
```

`serveError` would look something like this:

```go
func serveError(w http.ResponseWriter, r *http.Request, err error) {
        w.Header().Set("Content-Type", "application/json")
        
	switch v := err.(type) {
	case HTTPError:
		w.WriteHeader(v.Status)
		serveError(w, r, v.Cause)
        case v2.Errors:
		w.WriteHeader(http.StatusBadRequest)
		serveJSON(w, v) // or just serve it right here
	case v2.ErrorCode:
		// just a naked errorcode, wrap and recurse
		serveError(w, r, v2.Error{Code: v, Message: v.Descriptor().Message})
	default:
		// just cast to v2.Errors
		var errors v2.Errors
		errors.PushErr(v)
		serveError(w, r, v)
	}
}
```

This has the following benefits:

1. The API maintains the `error` type as the `error` type. Defining your own error type is almost always the wrong approach.
2. Type genericity is maintained in a single, dirty function, which can be used to dictate how all status codes are reported.
3. No matter what error is returned anywhere, we get a valid json error response with the right format.
4. The error doesn't need to be over-specified at the return site for handler functions.